src/ints/mouse.cpp

```diff
@@ -18,1549 +18,932 @@
-
-	// Draw Mousecursor
-	dataPos = addy * CURSORX;
-	for (y = y1; y <= y2; y++) {
-		uint16_t scMask = mouse.screenMask[addy + y - y1];
-		uint16_t cuMask = mouse.cursorMask[addy + y - y1];
-		if (addx1 > 0) {
-			scMask <<= addx1;
-			cuMask <<= addx1;
-			dataPos += addx1;
-		};
-		for (x = x1; x <= x2; x++) {
-			uint8_t pixel = 0;
-			// ScreenMask
-			if (scMask & HIGHESTBIT)
-				pixel = mouse.backData[dataPos];
-			scMask <<= 1;
-			// CursorMask
-			if (cuMask & HIGHESTBIT)
-				pixel = pixel ^ 0x0F;
-			cuMask <<= 1;
-			// Set Pixel
-			INT10_PutPixel(x, y, mouse.page, pixel);
-			dataPos++;
-		};
-		dataPos += addx2;
-	};
-	RestoreVgaRegisters();
-}
+} MouseEvent;
 
-static void CursorMoved(float xrel, float yrel, float x, float y, bool emulate)
+static class MouseQueue final {
+public:
+    MouseQueue();
+
+    void AddEvent(MouseEvent &event);
+    void FetchEvent(MouseEvent &event);
+    void ClearEventsDOS();
+    void StartTimerIfNeeded();
+    void Tick();
+
+    uint8_t start_delay_ps2     = 5; // for PS/2 events, in milliseconds
+    uint8_t start_delay_dos_btn = 1; // for DOS button events
+    uint8_t start_delay_dos_mov = 5; // for DOS move/wheel events
+
+private:
+    MouseQueue(const MouseQueue &)            = delete;
+    MouseQueue &operator=(const MouseQueue &) = delete;
+
+    void AggregateEventsDOS(MouseEvent &event);
+    MouseEvent &PopEventBtn();
+    void UpdateDelayCounters();
+
+    std::array<MouseEvent, 8> events = {}; // a modulo queue of events
+    uint8_t idx_first                = 0;  // index of the first event
+    uint8_t num_events               = 0;  // number of events in the queue
+
+    MouseButtons12S last_buttons_12S = 0; // last buttons reported
+
+    bool queue_overflow    = false;
+    bool timer_in_progress = false;
+
+    // Time in milliseconds which has to elapse before event can take place
+    uint8_t delay_ps2     = 0;
+    uint8_t delay_dos_btn = 0; // for DOS button events
+    uint8_t delay_dos_mov = 0; // for DOS move/wheel events
+
+    // Events for which a flag is enough to store them
+    bool event_ps2       = false;
+    bool event_dos_moved = false;
+    bool event_dos_wheel = false;
+
+    bool prefer_ps2 = false; // true = next time prefer PS/2 event
+
+    uint32_t ticks_start = 0; // PIC_Ticks value when timer starts
+
+    // Helper masks for aggregating DOS button events
+    static constexpr uint8_t aggr_mask_pressed =
+            static_cast<uint8_t>(MouseEventId::PressedLeft)  |
+            static_cast<uint8_t>(MouseEventId::PressedRight) |
+            static_cast<uint8_t>(MouseEventId::PressedMiddle);
+    static constexpr uint8_t aggr_mask_released =
+            static_cast<uint8_t>(MouseEventId::ReleasedLeft)  |
+            static_cast<uint8_t>(MouseEventId::ReleasedRight) |
+            static_cast<uint8_t>(MouseEventId::ReleasedMiddle);
+
+    // Helpers to check if there are events in the queue
+    inline bool HasEventDOSmov() const
```

Comment:
The casing is quite wonky in all these methods... TBH, I'd just use `Dos` instead of `DOS` in all the method names. Not reallly a fan of using uppercased acronyms in names at all...

`HasEventDosMoved`
`HasEventDosButton`
...